File: servo/components/style/values/computed/ui.rs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

//! Computed values for UI properties

use values::computed::color::Color;
use values::computed::url::ComputedImageUrl;
use values::computed::Number;
use values::generics::ui as generics;
use values::{Auto, Either};

pub use values::specified::ui::MozForceBrokenImageIcon;

/// auto | <color>
pub type ColorOrAuto = Either<Color, Auto>;

/// A computed value for the `cursor` property.
pub type Cursor = generics::Cursor<CursorImage>;

/// A computed value for item of `image cursors`.
pub type CursorImage = generics::CursorImage<ComputedImageUrl, Number>;

/// A computed value for `scrollbar-color` property.
pub type ScrollbarColor = generics::ScrollbarColor<Color>;

File: devtools/client/debugger/src/utils/ui.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

/* Checks to see if the root element is available and
 * if the element is visible. We check the width of the element
 * because it is more reliable than either checking a focus state or
 * the visibleState or hidden property.
 */
export function isVisible() {
  const el = document.querySelector("#mount");
  return !!(el && el.getBoundingClientRect().width > 0);
}

/* Gets the line numbers width in the code editor
 */
export function getLineNumberWidth(editor) {
  const { gutters } = editor.display;
  const lineNumbers = gutters.querySelector(".CodeMirror-linenumbers");
  return lineNumbers?.clientWidth;
}

/**
 * Forces the breakpoint gutter to be the same size as the line
 * numbers gutter. Editor CSS will absolutely position the gutter
 * beneath the line numbers. This makes it easy to be flexible with
 * how we overlay breakpoints.
 */
export function resizeBreakpointGutter(editor) {
  const { gutters } = editor.display;
  const breakpoints = gutters.querySelector(".breakpoints");
  if (breakpoints) {
    breakpoints.style.width = `${getLineNumberWidth(editor)}px`;
  }
}

/**
 * Updates CSS variables to reflect the current dimensions of the editor
 * and its gutter.
 * This is used to force the left toggle button in source header to be the same size
 * as the line numbers gutter, and to properly position the conditional panel.
 */
export function updateEditorSizeCssVariables(cmEditorEl) {
  document.documentElement.style.setProperty(
    "--cm-gutter-width",
    `${cmEditorEl.querySelector(".cm-gutters").clientWidth}px`
  );
  document.documentElement.style.setProperty(
    "--cm-editor-scroller-width",
    // We want the scroller width *without* a possible vertical scrollbar
    `${cmEditorEl.querySelector(".cm-scroller").clientWidth}px`
  );
}
